myia/prim/type_inferrers.py

```diff
@@ -532,3 +533,35 @@ def chk(data_v, item_v):
     xref = TransformedReference(track.engine, getelement, lst)
     res_elem_t = await track.assert_same(fn_t(xref, xref), dflt)
     return res_elem_t
+
+
+@type_inferrer(P.embed, nargs=1)
+async def infer_type_embed(track, x):
+    """Infer the return type of embed."""
+    return SymbolicKeyType
+
+
+@type_inferrer(P.env_setitem, nargs=3)
+async def infer_type_env_setitem(track, env, key, x):
+    """Infer the return type of env_setitem."""
+    await track.check(EnvType, env)
+    await track.check(SymbolicKeyType, key)
+    key_v = await key['value']
+    await track.assert_same(track.engine.vref(key_v.inferred), x)
+    return EnvType
+
+
+@type_inferrer(P.env_getitem, nargs=3)
+async def infer_type_env_getitem(track, env, key, default):
+    """Infer the return type of env_getitem."""
+    await track.check(EnvType, env)
+    await track.check(SymbolicKeyType, key)
+    key_v = await key['value']
```

Comment:
What happens if you get ANYTHING here?

Reply:
I suppose that could happen if you retrieved it from a data structure. It shouldn't happen under normal circumstances, though, so I'll just throw in an assert.